tiri/src/lib.rs

```diff
@@ -11,28 +11,131 @@
+/// Each local variable in the interpreter maps to one `Value`.
+/// This pretty much emulates Miri's `Operand` type.
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum Value {
+    Immediate(Immediate),
+    _Indirect, // FIXME Not implemented.
+}
+
+impl Value {
+    /// A helper to create a raw scalar value of the given value and size.
+    #[cfg(test)]
+    fn raw_scalar(data: u128, size: u8) -> Self {
+        Value::Immediate(Immediate::Scalar(Scalar::Raw { data, size }))
+    }
+
+    /// A helper to cast a value to a Boolean, if appropriate.
+    fn as_bool(&self) -> bool {
+        match self {
+            Value::Immediate(i) => i.as_bool(),
+            _ => panic!("Invalid cast from Value to bool"),
+        }
+    }
+
+    /// A helper to get the u128 value from inside a raw immediate.
+    fn as_u128(&self) -> u128 {
+        match self {
+            Value::Immediate(Immediate::Scalar(Scalar::Raw { data, .. })) => *data,
+            _ => panic!("Invalid cast from Value to u128"),
+        }
+    }
+}
+
+impl From<bool> for Value {
+    fn from(b: bool) -> Self {
+        Self::Immediate(Immediate::from(b))
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum Immediate {
+    Scalar(Scalar),
+    ScalarPair(Scalar, Scalar),
+}
+
+impl From<bool> for Immediate {
+    fn from(b: bool) -> Self {
+        Self::Scalar(Scalar::from(b))
+    }
+}
+
+impl Immediate {
+    fn as_bool(&self) -> bool {
+        match self {
+            // In the case of a `ScalarPair` we assume we are casting to bool to inspect the result
+            // of a checked operation, hence we cast the second element.
+            Immediate::Scalar(s) | Immediate::ScalarPair(_, s) => s.as_bool(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum Scalar {
+    Raw { data: u128, size: u8 },
+    _Ptr, // FIXME Not implemented.
+}
+
+impl From<bool> for Scalar {
+    fn from(b: bool) -> Self {
+        if b {
+            Scalar::Raw { data: 1, size: 1 }
+        } else {
+            Scalar::Raw { data: 0, size: 1 }
+        }
+    }
+}
+
+impl Scalar {
+    fn as_bool(&self) -> bool {
+        let data = match self {
+            Scalar::Raw { data, .. } => data,
+            Scalar::_Ptr => panic!("Invalid cast from Scalar to bool"),
+        };
+        *data != 0
+    }
+}
 
 /// Mutable interpreter state.
-struct InterpState {
+pub struct InterpState {
     /// The next position in the trace to interpret.
-    trace_pos: usize,
+    pc: usize,
     /// Local variable store.
-    _locals: Vec<Local>,
+    locals: HashMap<u32, Value>,
+    /// Set true upon aborting the execution of a trace.
+    abort: bool,
 }
 
 impl InterpState {
-    fn new() -> Self {
+    pub fn new() -> Self {
         Self {
-            trace_pos: 0,
-            _locals: Vec::new(),
+            pc: 0,
+            locals: HashMap::new(),
+            abort: false,
+        }
+    }
+}
+
+impl InterpState {
+    fn store_local(&mut self, local: &Local, val: Value) {
```

Comment:
Maybe implement Copy for Local and pass it by val instead. That will be faster when this is not inlined.

Reply:
I believe you, but why would it be faster?

Reply:
Because otherwise calling `.idx()` requires accessing the memory behind a pointer, while using `Local` without `&` it is possible (and likely) for the compiler to put the `Local` in a register.

Reply:
OK.

(we are likely to make `Local` a plain integer in the future anyway, so we will in fact copy soon)

Reply:
I thought `Local` was already just a newtype around an integer.